### PR TITLE
Etw service

### PR DESCRIPTION
**Pull request by bmcder02**
Changed watch_etw to support multiple providers under a single session. Should address issue #2944 .

### CODE REVIEW

File: vql/windows/etw/watcher.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+		scope.Log("Unregistering %v\n", handle.guid)
+
+		self.mu.Lock()
+		sessionContext, pres := self.sessions[session_name]
+		self.mu.Unlock()
+		if !pres {
+			return
+		}
+
+		sessionContext.mu.Lock()
```
Comment (scudette) [inline vql/windows/etw/watcher.go:74]: Move this method to the context itself.

File: vql/windows/etw/watcher.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+		defer sessionContext.mu.Unlock()
+
+		new_reg := make([]*Handle, len(sessionContext.registrations))
+		for _, handle := range sessionContext.registrations {
+			if handle.guid != wGuid {
```
Comment (scudette) [inline vql/windows/etw/watcher.go:79]: we probably need to manage keywords better here

File: vql/windows/etw/watcher.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+		// Create a scope with a completely different lifespan since
+		// it may outlive this query (if another query starts watching
+		// the same file). The query will inherit the same ACL
+		// manager, log manager etc but this is usually fine as there
+		// are not different ACLs managers on the client side.
+		manager := &repository.RepositoryManager{}
+		builder := services.ScopeBuilderFromScope(scope)
+		subscope := manager.BuildScope(builder)
+		frequency := vql_subsystem.GetIntFromRow(
+			scope, scope, constants.EVTX_FREQUENCY)
```
Comment (scudette) [inline vql/windows/etw/watcher.go:115]: we dont need this

File: vql/windows/etw/watcher.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+	ctx context.Context, scope vfilter.Scope, sessionContext *SessionContext,
+	key string, frequency uint64, cancel context.CancelFunc) {
+
+	defer utils.CheckForPanic("StartMonitoring")
+
+	// By default check every 15 seconds. Event logs are not flushed
+	// that often so checking more frequently does not help much.
+	if frequency == 0 {
+		frequency = 15
+	}
```
Comment (scudette) [inline vql/windows/etw/watcher.go:150]: dont need this

File: vql/windows/etw/watcher.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+		if err == nil {
+			event.Set("EventData", data)
+		}
+
+		if !sessionContext.status {
+			return
+		}
+
+		sessionContext.mu.Lock()
+		handles := sessionContext.registrations
```
Comment (scudette) [inline vql/windows/etw/watcher.go:167]: Might be worth taking a copy here just in case

File: vql/windows/etw/watcher.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+func (self *SessionContext) Close() {
+	self.mu.Lock()
+	defer self.mu.Unlock()
+	self.status = false
+
+	// Empty the registrations and close the session.
+	self.registrations = []*Handle{}
+	self.session.Close()
+	delete(GlobalEventTraceService.sessions, self.name)
```
Comment (scudette) [inline vql/windows/etw/watcher.go:215]: move to a method - this is not thread safe

**Review by scudette — CHANGES_REQUESTED** [review]

File: vql/windows/etw/watcher.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+	// Check if we already have a session for this provider.
+	sessionContext, pres := self.sessions[session_name]
+	if !pres {
+		var err error
+		sessionContext, err = self.NewSessionContext(session_name, scope)
+		if err != nil {
+			return cancel, nil, err
+		}
+
+		// err = sessionContext.UpdateSession(ctx, scope, wGuid, any_keyword, all_keyword, level)
```
Comment (scudette) [inline vql/windows/etw/watcher.go:78]: clean up?

File: vql/windows/etw/watcher.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+	err := sessionContext.UpdateSession(
+		ctx, scope, wGuid, any_keyword,
+		all_keyword, level)
+	if err != nil {
+		return cancel, output_chan, err
+	}
+
+	scope.Log("Registering watcher for %v", provider_guid)
+	sessionContext.mu.Lock()
```
Comment (scudette) [inline vql/windows/etw/watcher.go:105]: move to a method on session

File: vql/windows/etw/watcher.go [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+		data, err := e.EventProperties()
+		if err == nil {
+			event.Set("EventData", data)
+		}
+
+		if !sessionContext.status {
+			return
+		}
+
+		sessionContext.mu.Lock()
```
Comment (scudette) [inline vql/windows/etw/watcher.go:135]: move to method

**Review by scudette — CHANGES_REQUESTED** [review]

**Comment by bmcder02** [comment]
Waiting on ETW library change to be merged before Windows Test will pass.

**Pull request merged**